src/packaging/wheelfile.py

```diff
@@ -0,0 +1,544 @@
+from __future__ import annotations
+
+__all__ = [
+    "WheelMetadata",
+    "WheelRecordEntry",
+    "WheelContentElement",
+    "WheelError",
+    "WheelArchiveFile",
+    "WheelReader",
+    "write_wheelfile",
+    "WheelWriter",
+]
+
+import csv
+import hashlib
+import os.path
+import stat
+import time
+from base64 import urlsafe_b64decode, urlsafe_b64encode
+from collections import OrderedDict
+from collections.abc import Iterable, Iterator
+from contextlib import ExitStack
+from datetime import datetime, timezone
+from email.message import Message
+from email.policy import EmailPolicy
+from io import BytesIO, StringIO, UnsupportedOperation
+from os import PathLike
+from pathlib import Path, PurePath
+from types import TracebackType
+from typing import IO, NamedTuple
+from zipfile import ZIP_DEFLATED, ZIP_STORED, ZipFile, ZipInfo
+
+from .tags import Tag
+from .utils import (
+    BuildTag,
+    InvalidWheelFilename,
+    NormalizedName,
+    parse_wheel_filename,
+)
+from .version import Version
+
+_exclude_filenames = ("RECORD", "RECORD.jws", "RECORD.p7s")
+_default_timestamp = datetime(1980, 1, 1, tzinfo=timezone.utc)
+_email_policy = EmailPolicy(max_line_length=0, mangle_from_=False, utf8=True)
+
+
+class WheelMetadata(NamedTuple):
+    name: NormalizedName
+    version: Version
+    build_tag: BuildTag
+    tags: frozenset[Tag]
+
+    @classmethod
+    def from_filename(cls, fname: str) -> WheelMetadata:
```

Comment:
Would it make sense to have the other direction, too? A `to_filename` method? Or even just a `__str__` that would produce the filename? I had a method like this:

```python
@property
def name_ver(self) -> str:
    name = self.metadata.canonical_name.replace("-", "_")
    # replace - with _ as a local version separator
    version = str(self.metadata.version).replace("-", "_")
    return f"{name}-{version}"

@property
def basename(self) -> str:
    pyver = ".".join(sorted({t.interpreter for t in self.tags}))
    abi = ".".join(sorted({t.abi for t in self.tags}))
    arch = ".".join(sorted({t.platform for t in self.tags}))
    optbuildver = (
        [self.wheel_metadata.build_tag] if self.wheel_metadata.build_tag else []
    )
    return "-".join([self.name_ver, *optbuildver, pyver, abi, arch])
```

Reply:
> Why isn't this BuildTag | None, by the way? I guess it supports an empty tuple. Though then it seems make_wheel_filename could default to an empty tuple instead of supporting None.

Can you clarify what you're talking about here?